File: frontend/src/lib/progress.ts
import type { ExtractionPhase, ProgressUpdate } from "./types";

/**
 * One progress vocabulary for every extraction mode.
 *
 * The backend reports which phase it is in and how many units it has finished;
 * this file is the only place that turns that into a percentage and a label, so
 * the bar reads the same whichever pipeline ran underneath.
 */

export const UPLOADING_LABEL = "Enviando arquivo...";

export const PHASE_LABELS: Record<ExtractionPhase, string> = {
  queued: "Aguardando processamento...",
  detecting: "Identificando melhor abordagem...",
  processing: "Processando arquivo...",
  building: "Montando planilha...",
};

// Each phase owns a slice of the bar. Only processing has units to count, so it
// gets nearly all of it; the rest hold their edge and only change the label.
export const PHASE_RANGES: Record<ExtractionPhase, [number, number]> = {
  queued: [0, 0],
  detecting: [0, 0],
  processing: [0, 99],
  building: [99, 100],
};

/** Percentage for a frame, or null when it carries no phase to place it in. */
export function phasePercent({ phase, chunk, total }: ProgressUpdate): number | null {
  if (!phase || !(phase in PHASE_RANGES)) return null;
  const [start, end] = PHASE_RANGES[phase];
  const units = typeof total === "number" && total > 0 ? total : 1;
  const finished = typeof chunk === "number" && Number.isFinite(chunk) ? chunk : 0;
  const ratio = Math.min(1, Math.max(0, finished / units));
  return Math.round(start + ratio * (end - start));
}

export function phaseLabel(phase: ExtractionPhase | undefined): string | null {
  return phase && phase in PHASE_LABELS ? PHASE_LABELS[phase] : null;
}

File: backend/services/progress.py
"""Progress frames shared by every extraction pipeline.

Every mode reports the same server-side phases, so the UI can render one bar
without knowing which pipeline ran (uploading is earlier, and client-side):

    queued      the single processing slot is busy, this document has not begun
    detecting   choosing an approach - native text, local OCR, or a paid model
    processing  the real work, counted in finished units
    building    assembling the spreadsheet once extraction is done

`done`/`total` always count *finished* units, never positions, because chunks
run concurrently and complete out of order. A stage with nothing to count
reports a single unit, so the bar never has to special-case it.

`payload` is the frame as a dict, for the polled job endpoint; `progress` wraps
the same dict as an SSE line. Both carry `phase`, which is what the UI reads -
a frame without one cannot be placed on the bar at all.
"""
from __future__ import annotations

import json

QUEUED = "queued"
DETECTING = "detecting"
PROCESSING = "processing"
BUILDING = "building"

# One wording per phase, whatever the pipeline underneath. The stage names the
# pipeline actually went through (pdfplumber, tesseract, gemini) stay in `step`
# and in the logs: they mean nothing to the person watching the bar.
QUEUED_LABEL = "Aguardando processamento..."
DETECTING_LABEL = "Identificando melhor abordagem..."
PROCESSING_LABEL = "Processando arquivo..."
BUILDING_LABEL = "Montando planilha..."


def payload(
    phase: str,
    message: str,
    done: int = 0,
    total: int = 1,
    step: str | None = None,
) -> dict:
    """One progress frame as a dict. `step` names the stage, for logs only."""
    frame: dict = {
        "type": "progress",
        "phase": phase,
        "message": message,
        "chunk": done,
        "total": max(1, total),
    }
    if step:
        frame["step"] = step
    return frame


def progress(
    phase: str,
    message: str,
    done: int = 0,
    total: int = 1,
    step: str | None = None,
) -> str:
    """The same frame as an SSE line."""
    return "data: " + json.dumps(payload(phase, message, done, total, step), ensure_ascii=False) + "\n\n"


def queued_payload() -> dict:
    """For the polled endpoint, where a queued document has no live stream."""
    return payload(QUEUED, QUEUED_LABEL)


def queued(step: str | None = None) -> str:
    return progress(QUEUED, QUEUED_LABEL, 0, 1, step=step)


def detecting(step: str | None = None) -> str:
    return progress(DETECTING, DETECTING_LABEL, 0, 1, step=step)


def processing(done: int = 0, total: int = 1, step: str | None = None) -> str:
    return progress(PROCESSING, PROCESSING_LABEL, done, total, step=step)


def building(step: str | None = None) -> str:
    return progress(BUILDING, BUILDING_LABEL, 0, 1, step=step)


def keep_alive() -> str:
    """A comment frame: holds the connection open without moving the bar."""
    return ": keep-alive\n\n"
